File: 0206-reverse-linked-list/0206-reverse-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
    var prev, curr, next *ListNode
    prev, curr = nil, head
    for curr != nil {
        next = curr.Next
        curr.Next = prev
        prev, curr = curr, next
    }

    return prev
}

File: 0206-reverse-linked-list/0206-reverse-linked-list.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @return {ListNode}
def reverse_list(head)
    return if head.nil?
    return head if head.next.nil?

    stack = []

    curr = head
    while curr
        stack.push(curr)
        curr = curr.next
    end

    head = stack.pop
    curr = head

    while !stack.empty?
        node = stack.pop
        curr.next = node
        curr = node
    end
    curr.next = nil

    head
end